Adds support for Section target reports

sergiorg-hpc opened the pull request:
This commit adds support to enable reports for both a `Target Section`, or a `Target Compartment` with a `Section` inside. The behaviour is the same as in NEURON, providing the middle value in the first case, while providing all the values in the second case with Compartment.

To summarize the changes, the solution extends the purpose of one boolean in the `report.conf` file to encode new types of targets, such as `axon`, `axonComp` (for `Compartments`), and so on. This guarantees file compatibility with previous versions of CoreNEURON. The new encoded type is retrieved as an `enum class` and used to determine the type of report.

More information, including instructions on how to request the reports, can be found in the ticket:
https://bbpteam.epfl.ch/project/issues/browse/CNEUR-362?focusedCommentId=131257&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-131257

CI_BRANCHES:NEURON_BRANCH=master

pramodk commented:
> Can one of the admins verify this patch?

I added sergiorg-hpc username to CI plans so that we won't have above message and CI will trigger all build plans.

alkino commented on coreneuron/io/reports/report_configuration_parser.cpp:
```diff
@@ -39,6 +39,22 @@
+    Apical = 4,
+    AxonComp = 5,
+    DendriteComp = 6,
+    ApicalComp = 7
```
I like when the last has a `comma` too ;)

alkino commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -155,6 +165,77 @@ VarsToReport ReportHandler::get_compartment_vars_to_report(const NrnThread& nt,
+            std::vector<VarWithMapping> to_report;
+            to_report.reserve(cell_mapping->size());
+
+            std::string name;
+            switch (type) {
+                case AxonReport:
+                case AxonCompReport:
+                    name = "axon";
+                    break;
+                case DendriteReport:
+                case DendriteCompReport:
+                    name = "dend";
+                    break;
+                case ApicalReport:
+                case ApicalCompReport:
+                    name = "apic";
+                    break;
+                default:
+                    name = "soma";
+            }
```
Maybe you can define a function along the enum for that.

Maybe soma should be the default.

pramodk commented on coreneuron/io/reports/nrnreport.hpp:
```diff
@@ -51,7 +51,9 @@ namespace coreneuron {
-enum ReportType { SomaReport, CompartmentReport, SynapseReport, IMembraneReport };
+enum ReportType { SomaReport, CompartmentReport, SynapseReport, IMembraneReport,
+                  AxonReport, DendriteReport, ApicalReport, AxonCompReport,
+                  DendriteCompReport, ApicalCompReport };
```
nitpick : as we have more report types, here we should add some docstrings explaining what is different between DendriteReport vs DendriteCompReport etc.

pramodk replied:
Is DendriteCompReport means all compartments of dendrite and DendriteReport means central compartment of dendrite?
In that case either we can rename it or define it with docs:

```
enum ReportType { 
    SomaReport,           // here my explanation
    CompartmentReport,    // another type explanation

```

sergiorg-hpc replied:
We have modified the `enum` to contain a `SectionReport` type, instead of mixing types and subtypes. This way, the new `SectionType` and the compartment property can define the purpose of the report.

pramodk commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -155,6 +165,77 @@ VarsToReport ReportHandler::get_compartment_vars_to_report(const NrnThread& nt,
+                case AxonReport:
+                case AxonCompReport:
+                    name = "axon";
+                    break;
```
if we are converting `type` to `name` here, when `get_section_vars_to_report()` is called, is it possible to pass just `name`?

sergiorg-hpc replied:
Agreed, but we need to do the same thing externally. The latest changes simplify this part too, I hope it is now better and more understandable.

pramodk commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -155,6 +165,77 @@ VarsToReport ReportHandler::get_compartment_vars_to_report(const NrnThread& nt,
+                        }
+                    } else {
+                        /** corresponding voltage in coreneuron voltage array */
+                        const auto idx = vec[vec.size() / 2];
```
just for safety, can you assert here vec.size is odd?

jorblancoa replied:
Even you mean? It is always odd...

pramodk replied:
Here we are finding `middle` compartment. To have middle compartment we want #compartments to be odd.

sergiorg-hpc replied:
@pramodk, would it work for you just with the `assert()` as it is right now?

pramodk commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -155,6 +165,77 @@ VarsToReport ReportHandler::get_compartment_vars_to_report(const NrnThread& nt,
+                    auto& vec = sm.second;
+
+                    // Get all the compartment values (otherwise, just middle point)
+                    if (type >= AxonCompReport) {
```
semantically this comparison `type >= AxonCompReport` confusing. There is no specific order. Could this be done differently?

sergiorg-hpc replied:
This is no longer needed with the latest changes, but agreed! It could also introduce errors if someone else changes the order of the `enum` values.

pramodk reviewed (commented):
some minor comments/feedback.

alexsavulescu commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -155,6 +165,77 @@ VarsToReport ReportHandler::get_compartment_vars_to_report(const NrnThread& nt,
+                        /** corresponding voltage in coreneuron voltage array */
+                        const auto idx = vec[vec.size() / 2];
+                        double* variable = report_variable + idx;
+                        to_report.push_back(VarWithMapping(compartment_id, variable));
```
`emplace_back`

jorblancoa commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -155,6 +165,77 @@ VarsToReport ReportHandler::get_compartment_vars_to_report(const NrnThread& nt,
+                        for (const auto& idx : vec) {
+                            /** corresponding voltage in coreneuron voltage array */
+                            double* variable = report_variable + idx;
+                            to_report.push_back(VarWithMapping(compartment_id, variable));
```
This was copied from the other functions, I will create a PR for changing this in all the reporting functions.

jorblancoa commented on coreneuron/io/reports/report_configuration_parser.cpp:
```diff
@@ -96,10 +112,33 @@ std::vector<ReportConfiguration> create_report_configurations(const char* conf_f
-            else
-                report.type = CompartmentReport;
+            } else {
+                switch (target_type) {
```
It is not exactly 1to1. We have added another enum that shows more clearly the differences.

sergiorg-hpc commented:
@jorblancoa and I have introduced several changes that addresses the suggestions of the review, thank you so much for the insight!

To clarify, we have tested again generating all the reports on a small test case, and compared the result with NEURON. This is just to confirm that the latest changes didn't introduce any errors.

sergiorg-hpc commented:
Just to clarify, the CI is not passing because of the NEURON version issue (i.e., the branch was updated recently to `master`). However, we relaunched the CI from the `blueconfigs` PR and it works perfectly fine:

https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.SimulationStack/detail/hpc.SimulationStack/2687/pipeline
https://bbpcode.epfl.ch/code/#/c/51354/

alexsavulescu commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -155,6 +161,73 @@ VarsToReport ReportHandler::get_compartment_vars_to_report(const NrnThread& nt,
+                                                       SectionType section_type,
+                                                       bool all_compartments) const {
+    VarsToReport vars_to_report;
+    const auto section_type_str = getSectionTypeStr(section_type);
```
bind temporary to const reference
```suggestion
    const auto& section_type_str = getSectionTypeStr(section_type);
```

alexsavulescu commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -155,6 +161,73 @@ VarsToReport ReportHandler::get_compartment_vars_to_report(const NrnThread& nt,
+            return "axon";
+        case Dendrite:
+            return "dend";
+        default:
+            return "apic";
```
We can make this future proof:
```suggestion
        case Apical:
            return "apic";
        default:        
            std::cerr << "SectionType not handled in getSectionTypeStr" << std::endl;
            nrn_abort(1);
```

pramodk replied:
👍

pramodk commented on coreneuron/io/reports/report_configuration_parser.cpp:
```diff
@@ -39,6 +39,23 @@
+/*
+ * Defines the type of target, as per the following syntax:
+ *   0=Compartment, 1=Cell/Soma, Section { 2=Axon, 3=Dendrite, 4=Apical }
+ * The "Comp" variations are compartment-based (all values, not middle only)
```
```suggestion
 * The "Comp" variations are compartment-based (all segments, not middle only)
```

pramodk commented on coreneuron/io/reports/report_configuration_parser.cpp:
```diff
@@ -96,10 +113,45 @@ std::vector<ReportConfiguration> create_report_configurations(const char* conf_f
+                        report.section_all_compartments = true;
+                        break;
+                    default:
+                        report.type = CompartmentReport;
```
Minor  - Is there possibility to avoid `default` here? (like Alex suggested. The reason is to avoid surprise like what Sirio had when he specified Section target type)

sergiorg-hpc replied:
Agreed. Technically, we already filter most of the errors from the Neurodamus-side, but just in case, I completely agree.

pramodk reviewed (commented):
2 minor comments otherwise LGTM

alexsavulescu commented on coreneuron/io/reports/report_configuration_parser.cpp:
```diff
@@ -96,14 +113,53 @@ std::vector<ReportConfiguration> create_report_configurations(const char* conf_f
+                        break;
+                    default:
+                        std::cerr << "Report error: unsupported target type" << std::endl;
+                        abort();
```
we should be as consistent as possible
```suggestion
                        nrn_abort(1);
```

sergiorg-hpc replied:
I agree... But I used `abort()` because there are other places in that file using it, and not `nrn_abort()`.

alexsavulescu commented on coreneuron/io/reports/report_configuration_parser.cpp:
```diff
@@ -96,14 +113,53 @@ std::vector<ReportConfiguration> create_report_configurations(const char* conf_f
             report.type = SynapseReport;
         } else {
-            std::cerr << "Report error: unsupported type " << report.type_str << "\n";
+            std::cerr << "Report error: unsupported type " << report.type_str << std::endl;
```
same for the `abort()` here afterwards

sergiorg-hpc replied:
See my previous comment.

alexsavulescu reviewed (commented):
LGTM

alexsavulescu commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -155,6 +161,76 @@ VarsToReport ReportHandler::get_compartment_vars_to_report(const NrnThread& nt,
+                            to_report.push_back(VarWithMapping(compartment_id, variable));
+                        }
+                    } else {
+                        assert(vec.size() % 2);
```
I was thinking after our last discussion, maybe here we should have:
 ```suggestion
                        nrn_assert(vec.size() % 2);
```

iomaganaris commented:
please retest

pramodk commented:
So the conclusion is NEURON_BRANCH is somehow not working in this case?

iomaganaris commented:
> So the conclusion is NEURON_BRANCH is somehow not working in this case?

Yes, there is some issue with the multiline description and the way the NEURON_BRANCH is filtered in the CI. I opened a PR in the `blueconfigs` repo to address the issue. It should be finished today

iomaganaris commented:
please retest

sergiorg-hpc commented:
please retest

sergiorg-hpc commented:
Hi there!

Thanks to @iomaganaris' great work, we have the CI back on track and now it can be confirmed that the PR changes passes it. Nonetheless, the following PR is the only one that evaluates the support for `Section` targets inside CoreNEURON:
https://bbpcode.epfl.ch/code/#/c/51354

Is there any additional changes you would like to integrate, or can we merge this already? If so, can someone approve the PR, please?

sergiorg-hpc commented:
Thank you everyone for the help and insight!
